File: Adhiti_Shaili/Day 13/Subsets 2.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> ans;
    void rec(vector<int>& nums,vector<int>& v,int k,int s)
    {
        if(v.size()==k)
        {
            ans.push_back(v);
            return;           
        }
        
        for(int i=s;i<nums.size();i++)
        {
            v.push_back(nums[i]);
            rec(nums,v,k,i+1);
            v.pop_back();
            
            while(i<nums.size()-1&&nums[i]==nums[i+1])
            {
                i++;
            }
        }
    }
    
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        int n=nums.size();
        sort(nums.begin(),nums.end());
        for(int i=0;i<=n;i++)
        {
           vector<int> v;
           rec(nums,v,i,0);
        }
        return ans;
        
    }
};

File: parnamondal/Week - 3/Day - 1/Subsets 2.py
# Time Complexity - O(N*2to the power of N)
# Space Complexity  - O(N*2to the power of N)

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        n=len(nums)
        nums.sort()
        def backtrack(first=0,curr=[]):
            if len(curr)==k:
                if curr not in output:
                    output.append(curr[:])
                    return 
            for i in range (first,n):
                curr.append(nums[i])
                backtrack(i+1,curr)
                curr.pop()
        output=[]
        for k in range (0,n+1):
            backtrack()
        return output

        